llumnix/llm_engine_manager.py

```diff
@@ -250,28 +256,110 @@ async def _migrate(self) -> None:
             logger.error("unexpected exception occurs: {}".format(e))
             logger.error("exception traceback: {}".format(traceback.format_exc()))
 
+    async def rebuild_migrate_backend(self) -> None:
+        # Wait for all instances to finish migration
+        while any(self.instance_migrating.values()):
+            await asyncio.sleep(0.1)
+
+        # During rebuilding migrate backend, disable migrate
+        origin_config = self.enable_migration
+        self.enable_migration = False
+
+        async def run_task(alive_instances: List[str], task_name: str, *args, **kwargs):
+            tasks = []
+            for instance_name in alive_instances:
+                llumlet_handle = self.instances[instance_name]
+                tasks.append(llumlet_handle.execute_engine_method.remote("_run_workers", task_name, *args, **kwargs))
+
+            rets = await asyncio.gather(*tasks, return_exceptions=True)
+
+            dead_instances = set()
+            for instance_name, ret in zip(alive_instances, rets):
+                if isinstance(ret, ray.exceptions.RayActorError):
+                    dead_instances.add(instance_name)
+
+            if len(dead_instances) > 0:
+                self.scale_down(dead_instances, rebuild_migrate_backend=False)
+                # clear gloo migrate backend intermediate state
+                ray.kill(ray.get_actor("gloo_queue", "llumnix"))
+
+            return dead_instances
+
+        alive_instances = sorted(self.instances.keys())
+        pending_task = self.pending_rebuild_migration_instances
+        group_name = None
+
+        while len(alive_instances) > 0 and self.pending_rebuild_migration_instances > 0:
+            dead_instances = set()
+            group_name = random_uuid()
+            instance_rank = {instance_id: index for index, instance_id in enumerate(alive_instances)}
+            dead_instances = dead_instances.union(await run_task(alive_instances, "rebuild_migration_backend",
+                                                                  instance_rank, group_name))
+
+            if len(dead_instances) == 0 and self.pending_rebuild_migration_instances == pending_task:
+                dead_instances = dead_instances.union(await run_task(alive_instances, "warmup"))
```

Comment:
I think inplace update is enough here, a_set.update(a_set).

Reply:
you are right. union will create a new object. And update is inplace